OpenRA.Mods.Common/Activities/UnloadCargo.cs

```diff
@@ -56,40 +59,42 @@ IEnumerable<CPos> BlockedExitCells(Actor passenger)
 		public override Activity Tick(Actor self)
 		{
 			cargo.Unloading = false;
-			if (IsCanceled || cargo.IsEmpty(self))
+			if (IsCanceled || !actorsToUnload.Any() || !cargo.Passengers.Intersect(actorsToUnload).Any())
 				return NextActivity;
 
 			if (cloak != null && cloak.Info.UncloakOnUnload)
 				cloak.Uncloak();
 
-			var actor = cargo.Peek(self);
-			var spawn = self.CenterPosition;
-
-			var exitSubCell = ChooseExitSubCell(actor);
-			if (exitSubCell == null)
+			foreach (var actor in actorsToUnload)
```

Comment:
You probably want to split the exiting over separate ticks, with a few ticks in between actors, so the ACP/building doesn't spawn 20 infantry at once (actually RA2 structures did that :| ).
RA2 transports, however, had delay between units exiting them, which made it look better (there is only one door, after all).

P.S.: I suggest you do mostly what the old code did, but instead of popping out of `cargo`, take actors from `actorsToUnload`.